src/language_features/code_lens.rs

```diff
@@ -32,10 +32,54 @@ pub fn text_document_code_lens(meta: EditorMeta, ctx: &mut Context) {
         partial_result_params: PartialResultParams::default(),
     };
     ctx.call::<CodeLensRequest, _>(meta, req_params, |ctx: &mut Context, meta, result| {
-        editor_code_lens(meta, result, ctx)
+        editor_code_lens(meta.clone(), result.clone(), ctx);
+        inlay_hints_from_editor_code_lens(meta, result.unwrap_or_default(), ctx)
     });
 }
 
+fn inlay_hints_from_editor_code_lens(
+    meta: EditorMeta,
+    mut lenses: Vec<CodeLens>,
+    ctx: &mut Context,
+) {
+    let document = match ctx.documents.get(&meta.buffile) {
+        Some(document) => document,
+        None => {
+            return;
+        }
+    };
+    lenses.sort_by_key(|lens| lens.range.start);
+
+    let inlay_hints_from_lenses = lenses
+        .iter()
+        .map(|lens| {
+            let label = lens.command.as_ref().map_or("", |v| &v.title);
+            if label.trim().len() > 0 {
+                let position =
+                    lsp_position_to_kakoune(&lens.range.start, &document.text, ctx.offset_encoding);
+                let line = position.line;
+                let col = position.column;
+                editor_quote(&format!(
+                    "{line}.{col}+0|{{InlayHint}}[\u{1F50E} {}] {{InlayHint}}",
+                    escape_tuple_element(label)
+                ))
+            } else {
+                String::new()
+            }
+        })
+        .join(" ");
+    let command = format!(
+        "set-option buffer lsp_inlay_hints {} {}",
```

Comment:
I believe LSP servers may send both `textDocument/inlayHint` and `textDocument/codeLens`, and the two are not necessarily related.
To allow this, I think we should give code lenses a dedicated replace-ranges highlighter with documented commands like `lsp-inlay-code-lenses-enable` (and use the name "inlay hints" only for code related to `textDocument/inlayHint`).

Reply:
Interesting point. I will explore this issue a bit more.

Reply:
I guess the alternative to my suggestion is to implement some kind of merging logic (like we do for our flag-lines highlighter, which is a combination of diagnostics and code lenses).

The question is if we want to expose a single option for enabling both, or separate options.
I'm fine with either direction.
At first, I was a bit apprehensive of using the term "inlay hints" for an umbrella option but it sounds like a reasonable choice.

Reply:
Long-term we might have to split them up.
But I think today there is no server that supports both inlay hints and code lenses (though I haven't checked at all), so making `lsp-inlay-hints-enable` activate both might just be fine..

Reply:
(I think the current version of this PR breaks rust-analyzer's inlay hints, both approaches would fix that).

Reply:
This TODO is still pending. Have a look at some other feedback items that I've addressed that I've made in the meanwhile. I will address this point in my next round of changes.

Reply:
Looks fine. The emoji thing is ugly but I'm sure we can remove that workaround eventually (not sure if it's necessary for current macOS).